### PR TITLE
Prepare for DIP3 operator reward payments and switch to array in getblocktemplate

### PR DESCRIPTION
**Pull request by codablock**
~This PR also includes the commits of #2215. I'll rebase and force-push after that one is merged. Only the last commit is actually part of this PR.~

This refactoring might look useless without the operator rewards from DIP3, but I'd like at least include the getblocktemplate change ASAP so that miners and pools operators can prepare themself for the upcoming changes.

### CODE REVIEW

File: src/masternode-payments.cpp [hunk 1]
```diff
@@ -205,7 +205,7 @@ bool IsBlockPayeeValid(const CTransaction& txNew, int nBlockHeight, CAmount bloc
     return true;
 }
 
-void FillBlockPayments(CMutableTransaction& txNew, int nBlockHeight, CAmount blockReward, CTxOut& txoutMasternodeRet, std::vector<CTxOut>& voutSuperblockRet)
+void FillBlockPayments(CMutableTransaction& txNew, int nBlockHeight, CAmount blockReward, std::vector<CTxOut>& txoutsMasternodeRet, std::vector<CTxOut>& voutSuperblockRet)
```
Comment (UdjinM6) [inline src/masternode-payments.cpp]: `std::vector<CTxOut>`s have `vout` prefix in code currently (`vout` in `CTransaction` and `voutSuperblockRet` in mnpayments/governance/blocktemplate), let's keep it consistent maybe i.e. s/`txoutsMasternodeRet`/`voutMasternodeRet`/ or (even more descriptive) s/`txoutsMasternodeRet`/`voutMasternodePaymentsRet`/ ?

File: src/masternode-payments.cpp [hunk 2]
```diff
@@ -277,19 +283,17 @@ bool CMasternodePayments::UpdateLastVote(const CMasternodePaymentVote& vote)
...
 
 /**
-*   FillBlockPayee
+*   GetMasternodeTxOuts
 *
 *   Get masternode payment tx output
 */
 
-bool CMasternodePayments::GetMasternodeTxOut(int nBlockHeight, CAmount blockReward, CTxOut &txoutMasternodeRet) const
+bool CMasternodePayments::GetMasternodeTxOuts(int nBlockHeight, CAmount blockReward, std::vector<CTxOut>& vecTxOutsMasternodeRet) const
```
Comment (UdjinM6) [inline src/masternode-payments.cpp]: same

File: src/miner.h [hunk 3]
```diff
@@ -30,7 +30,7 @@ struct CBlockTemplate
     CBlock block;
     std::vector<CAmount> vTxFees;
     std::vector<int64_t> vTxSigOps;
-    CTxOut txoutMasternode; // masternode payment
+    std::vector<CTxOut> txoutsMasternode; // masternode payment
```
Comment (UdjinM6) [inline src/miner.h]: same

File: src/masternode-payments.cpp [hunk 4]
```diff
@@ -504,12 +508,20 @@ bool CMasternodePaymentVote::Sign()
     return true;
 }
 
-bool CMasternodePayments::GetBlockPayee(int nBlockHeight, CScript& payeeRet) const
+bool CMasternodePayments::GetBlockTxOuts(int nBlockHeight, CAmount blockReward, std::vector<CTxOut>& vecTxOutsRet) const
```
Comment (UdjinM6) [inline src/masternode-payments.cpp]: same

File: src/masternode-payments.cpp [hunk 5]
```diff
@@ -523,11 +535,14 @@ bool CMasternodePayments::IsScheduled(const masternode_info_t& mnInfo, int nNotB
     CScript mnpayee;
     mnpayee = GetScriptForDestination(mnInfo.keyIDCollateralAddress);
 
-    CScript payee;
     for(int64_t h = nCachedBlockHeight; h <= nCachedBlockHeight + 8; h++){
         if(h == nNotBlockHeight) continue;
-        if(GetBlockPayee(h, payee) && mnpayee == payee) {
-            return true;
+        std::vector<CTxOut> txOuts;
```
Comment (UdjinM6) [inline src/masternode-payments.cpp]: same

File: src/masternode-payments.cpp [hunk 6]
```diff
@@ -691,9 +706,30 @@ std::string CMasternodePayments::GetRequiredPaymentsString(int nBlockHeight) con
 
 bool CMasternodePayments::IsTransactionValid(const CTransaction& txNew, int nBlockHeight, CAmount blockReward) const
 {
-    LOCK(cs_mapMasternodeBlocks);
-    const auto it = mapMasternodeBlocks.find(nBlockHeight);
-    return it == mapMasternodeBlocks.end() ? true : it->second.IsTransactionValid(txNew);
+    std::vector<CTxOut> masternodeTxOuts;
```
Comment (UdjinM6) [inline src/masternode-payments.cpp]: same

File: src/rpc/mining.cpp [hunk 7]
```diff
@@ -490,10 +492,10 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
     }
 
     // when enforcement is on we need information about a masternode payee or otherwise our block is going to be orphaned by the network
-    CScript payee;
+    std::vector<CTxOut> payees;
```
Comment (UdjinM6) [inline src/rpc/mining.cpp]: same

File: src/validation.cpp [hunk 8]
```diff
@@ -1948,13 +1948,13 @@ int32_t ComputeBlockVersion(const CBlockIndex* pindexPrev, const Consensus::Para
         ThresholdState state = VersionBitsState(pindexPrev, params, pos, versionbitscache);
         const struct BIP9DeploymentInfo& vbinfo = VersionBitsDeploymentInfo[pos];
         if (vbinfo.check_mn_protocol && state == THRESHOLD_STARTED && !fAssumeMasternodeIsUpgraded) {
-            CScript payee;
+            std::vector<CTxOut> blockTxOuts;;
```
Comment (UdjinM6) [inline src/validation.cpp]: same

**Comment by codablock** [comment]
Did the renaming, including cherry-picking https://github.com/dashpay/dash/pull/2216/commits/c6830f060f1fa7f8b172a611e0e8b2a89f8caa55 which you proposed in #2083 

I'd however prefer to not rename method names in this PR, as I'll already have enough effort when later rebasing DIP3 :see_no_evil:

File: src/validation.cpp [hunk 9]
```diff
@@ -1948,13 +1948,13 @@ int32_t ComputeBlockVersion(const CBlockIndex* pindexPrev, const Consensus::Para
...
-            CScript payee;
+            std::vector<CTxOut> voutMasternodePayments;
             masternode_info_t mnInfo;
-            if (!mnpayments.GetBlockPayee(pindexPrev->nHeight + 1, payee)) {
+            if (!mnpayments.GetBlockTxOuts(pindexPrev->nHeight + 1, 0, voutMasternodePayments)) {
                 // no votes for this block
                 continue;
             }
-            if (!mnodeman.GetMasternodeInfo(payee, mnInfo)) {
+            if (!mnodeman.GetMasternodeInfo(voutMasternodePayments[0].scriptPubKey, mnInfo)) {
```
Comment (UdjinM6) [inline src/validation.cpp]: This looks like a temporary shortcut thb which would make little sense after there are many MNs paid in one block. Should probably write this part correctly i.e. "continue _if one of paid masternodes_ is unknown" and "continue if one of masternodes is not upgraded" below.

Comment (codablock) [reply src/validation.cpp]: Actually, the idea behind voutMasternodePayments is not to pay multiple masternodes but to split the masternode reward into masternode payment and operator reward payment. Even in the case of masternode shares, a block will still only pay a single masternode but split the reward into multiple recipients.

I added a commit on top which loops through all txouts and if none of them is a masternode payment, it continues.

**Pull request merged**